### PR TITLE
[entropy_src/rtl] Aggregate RepCnt & RepCnts events

### PR DESCRIPTION
**Pull request by martin-lueker**
This commit changes the timing of the RepCnt & RepCnts failure pulses
to make it easier to predict alerts and total failure statistics.

The challenges of interpretting and predicting the statistics of these
tests prior to this commit are described in issue #14078 (where this fix
has been proposed).

When combined with the corresponding changes to the scoreboard, this
brings the pass rate to 100% for the entropy_src_rng test.

Signed-off-by: Martin Lueker-Boden <martin.lueker-boden@wdc.com>

### CODE REVIEW

**Review by mwbranstad — APPROVED** [review]

**Comment by mwbranstad** [comment]
Seems like the window concept was originally applied to REPCNT tests, but removed since it does not apply. It seems like its being added back, with modifications. Just trying to recall the history of this set of tests.

**Comment by martin-lueker** [comment]
> Seems like the window concept was originally applied to REPCNT tests, but removed since it does not apply. It seems like its being added back, with modifications. Just trying to recall the history of this set of tests.

The distinction is that the _data_ is pulled from across multiple windows.  Though the _reporting_ is lined up with the windows.

**Comment by mwbranstad** [comment]
Ah thanks for the refresher.

File: hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv
Comment (tjaychen) [inline hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:90]: is there any reason we can't use `rep_cnt_fail` to just set a flop to 1, and only clear it when various conditions match?

File: hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv
Comment (tjaychen) [inline hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:99]: should `clear_i` also cause this to go to 0?

**Review by tjaychen — COMMENTED** [review]
this lgtm, but just wondering if some of the logic can be simplified.

File: hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv
Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:99]: Really good catch.  I think it might, though I would have to run in simulation it to see if that causes any issues.

Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:99]: Thanks for taking a look.  This is a tough one as it requires a bit of nuance.   

`Active_i` in `entropy_src`  means, "the module is enabled, and entropy is flowing".   Clearing `active_i` halts the flow of entropy, and renders the current window null.  

`clear_i` means: "Start all the HT's over as we are just now activating the module and starting the HT."  For the purposes of this pulse, there is no practical difference though between including `clear_i` or not including it.    

That said we can leave clear out here, from a practical perspective.   From a theoretical perspective, one can ask "what if some one clears it in the middle of a window?"  That is a good question, but i'm not sure what the practical answer would be.  For now I'm inclined to not include clear.

File: hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv
Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:90]: Yes there is.  `rep_cnt_fail` as a counter output only gets updated (high or low) once per sample.  This causes a problem for the last sample in each window, because it stays high through the end of the previous pulse into the next one, and so depending on how you sample it it could be double counted it.  That's why we create a sample pulse immediately after the prim_counter is updated, to latch the data and make sure it is first registered in the same window where the failure occurred.

**Comment by martin-lueker** [comment]
> this lgtm, but just wondering if some of the logic can be simplified.

It can, though only if you allow the repcnt to register several pulses per window.

File: hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv
Comment (tjaychen) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:90]: o okay, so i guess your `fail_smoothed_d` is basically what i'm describing, and you OR it with `fail_sampled` to basically cover the last cycle problem in the window? that makes sense.

Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:90]: Well, almost.  `Fail_smoothed_d` is what you are describing, but before you flip that to HIGH need to mask `rep_cnt_fail` off to make sure that it doesn't get double counted, hence `fail_sampled`.   I wholly agree that this is more complicated than I would like, especially since this is the absolute last pulse we want to misplace.

After talking to Mark, I think scoreboarding can be made to work just with `fail_sampled` as the pulse output, but it would require a bit more creativity.

File: hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv
Comment (tjaychen) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:99]: @martin-lueker this part is a bit weird to me right now. It looks like the way it's integrated, `active_i` going low seems to always cause `clear_i` to pulse, so they effectively do the same thing.

However, since the `repcnt` purposely allows them the possibility to be different, it feels like the error tracking should do the same. Just hypothetically, if someone were ever to connect `active_i` and `clear_i` to different controlling signals, and did in fact clear in the middle, wouldn't we want the error to go away? Otherwise I think I think the intent would be clearer to not have both `active_i` and `clear_i` and just use `active_i`.

**Review by tjaychen — APPROVED** [review]
lgtm overall, with a question on `clear_i`.

File: hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv
Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:99]: Clear_i pulses when active goes high..

Comment (tjaychen) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:99]: oops, read that wrong, but i still feel like it should be included. Just in case this dependency is changed in the future.

Comment (tjaychen) [reply hw/ip/entropy_src/rtl/entropy_src_repcnts_ht.sv:99]: this doesn't need to block the PR though, just something to think about.

I think it's unlikely we'll change the behavior of `active_i` and `clear_i`...but...for completeness sake..it feels like anything (other than there not being an error condition) that can clear the count should probably also clear the latched error.

**Comment by msfschaffner** [comment]
kokoro rerun

**Comment by martin-lueker** [comment]
Given that @tjaychen had the same concern I did (that this more complex than it needs to be, especially since it tweaks a critical signal) I'm closing this PR.   Some small timing shifts are needed, but these have been moved to #14238.

**Pull request closed without merging**
